File: DashwoodCalendar/src/main/java/com/dashwood/dashwoodcalendar/handler/DateConverter.java
package com.dashwood.dashwoodcalendar.handler;


import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.GregorianCalendar;
import java.util.Locale;

import ir.huri.jcal.JalaliCalendar;

public class DateConverter {

    public static String persianDateToGregorianDate(String persianDate) {
        try {
            SimpleDateFormat format = new SimpleDateFormat(
                    "yyyy-MM-dd", Locale.US);
            Date date = format.parse(persianDate);
            if (date == null) {
                return "";
            }
            String persianFormatDate = format.format(date);
            String[] persianDates = persianFormatDate.split("-");
            JalaliCalendar jalaliCalendar = new JalaliCalendar(Integer.parseInt(persianDates[0]),
                    Integer.parseInt(persianDates[1]), Integer.parseInt(persianDates[2]));
            GregorianCalendar gregorianCalendar = jalaliCalendar.toGregorian();
            int day = gregorianCalendar.get(Calendar.DAY_OF_MONTH);
            String stringDay;
            if (day < 10) {
                stringDay = "0" + day;
            } else {
                stringDay = String.valueOf(day);
            }
            return stringDay + "-"
                    + (gregorianCalendar.get(Calendar.MONTH) + 1) + "-"
                    + gregorianCalendar.get(Calendar.YEAR);
        } catch (ParseException e) {
            e.printStackTrace();
        }
        return "";
    }

    public static String persianDateToGregorianDate(int year, int month, int day) {
        JalaliCalendar jalaliCalendar = new JalaliCalendar(year, month, day);
        GregorianCalendar gregorianCalendar = jalaliCalendar.toGregorian();
        int newDay = gregorianCalendar.get(Calendar.DAY_OF_MONTH);
        String stringDay;
        if (newDay < 10) {
            stringDay = "0" + newDay;
        } else {
            stringDay = String.valueOf(newDay);
        }
        return gregorianCalendar.get(Calendar.YEAR) + "-"
                + (gregorianCalendar.get(Calendar.MONTH) + 1) + "-"
                + stringDay;
    }


    public static String persianDateToGregorianDateFirstReturnYear(String persianDate) {
        try {
            SimpleDateFormat format = new SimpleDateFormat(
                    "yyyy-MM-dd", Locale.US);
            Date date = format.parse(persianDate);
            if (date == null) {
                return "";
            }
            String persianFormatDate = format.format(date);
            String[] persianDates = persianFormatDate.split("-");
            JalaliCalendar jalaliCalendar = new JalaliCalendar(Integer.parseInt(persianDates[0]),
                    Integer.parseInt(persianDates[1]), Integer.parseInt(persianDates[2]));
            GregorianCalendar gregorianCalendar = jalaliCalendar.toGregorian();
            int day = gregorianCalendar.get(Calendar.DAY_OF_MONTH);
            String stringDay;
            if (day < 10) {
                stringDay = "0" + day;
            } else {
                stringDay = String.valueOf(day);
            }
            return gregorianCalendar.get(Calendar.YEAR) + "-" +
                    (gregorianCalendar.get(Calendar.MONTH) + 1) + "-" + stringDay;
        } catch (ParseException e) {
            e.printStackTrace();
        }
        return "";
    }

    public static String gregorianToPersianWithMonthStringName(String gregorianDateValue) {
        try {
            SimpleDateFormat format = new SimpleDateFormat(
                    "yyyy-MM-dd", Locale.US);
            Date date = format.parse(gregorianDateValue);
            if (date == null) {
                return "";
            }
            String stringFormatData = format.format(date);
            String[] gregorianDate = stringFormatData.split("-");
            JalaliCalendar jalaliCalendar = new JalaliCalendar(new GregorianCalendar(Integer.parseInt(gregorianDate[0]),
                    Integer.parseInt(gregorianDate[1]) - 1, Integer.parseInt(gregorianDate[2])));
            int day = jalaliCalendar.getDay();
            String stringDay;
            if (day < 10) {
                stringDay = "0" + day;
            } else {
                stringDay = String.valueOf(day);
            }
            return stringDay + " "
                    + jalaliCalendar.getMonthString() + " " + jalaliCalendar.getYear();
        } catch (ParseException e) {
            e.printStackTrace();
        }
        return "";
    }

    public static String gregorianToPersian(String gregorianDateValue) {
        String[] dates = gregorianDateValue.split(" ");
        gregorianDateValue = dates[0];
        try {
            SimpleDateFormat format = new SimpleDateFormat(
                    "yyyy-MM-dd", Locale.US);
            Date date = format.parse(gregorianDateValue);
            if (date == null) {
                return "";
            }
            String stringFormatData = format.format(date);
            String[] gregorianDate = stringFormatData.split("-");
            JalaliCalendar jalaliCalendar = new JalaliCalendar(new GregorianCalendar(Integer.parseInt(gregorianDate[0]),
                    Integer.parseInt(gregorianDate[1]) - 1, Integer.parseInt(gregorianDate[2])));
            int day = jalaliCalendar.getDay();
            String stringDay;
            if (day < 10) {
                stringDay = "0" + day;
            } else {
                stringDay = String.valueOf(day);
            }
            return jalaliCalendar.getYear() + "-" + jalaliCalendar.getMonth() + "-" + stringDay;
        } catch (ParseException e) {
            e.printStackTrace();
        }
        return "";
    }

    public static String gregorianToPersian(int year, int month, int day) {

        JalaliCalendar jalaliCalendar = new JalaliCalendar(new GregorianCalendar(year, month, day));
        int newDay = jalaliCalendar.getDay();
        String stringDay;
        if (newDay < 10) {
            stringDay = "0" + newDay;
        } else {
            stringDay = String.valueOf(newDay);
        }
        return jalaliCalendar.getYear() + "-" + jalaliCalendar.getMonth() + "-" + stringDay;
    }

    public static int getMonthLength(String gregorianDateValue) {
        String[] dates = gregorianDateValue.split(" ");
        gregorianDateValue = dates[0];
        try {
            SimpleDateFormat format = new SimpleDateFormat(
                    "yyyy-MM-dd", Locale.US);
            Date date = format.parse(gregorianDateValue);
            if (date == null) {
                return 0;
            }
            String stringFormatData = format.format(date);
            String[] gregorianDate = stringFormatData.split("-");
            JalaliCalendar jalaliCalendar = new JalaliCalendar(new GregorianCalendar(Integer.parseInt(gregorianDate[0]),
                    Integer.parseInt(gregorianDate[1]) - 1, Integer.parseInt(gregorianDate[2])));
            return jalaliCalendar.getMonthLength();
        } catch (ParseException e) {
            e.printStackTrace();
        }
        return 0;
    }

    public static String getMonthStartWeekName(String gregorianDateValue) {
        String[] dates = gregorianDateValue.split(" ");
        gregorianDateValue = dates[0];
        try {
            SimpleDateFormat format = new SimpleDateFormat(
                    "yyyy-MM-dd", Locale.US);
            Date date = format.parse(gregorianDateValue);
            if (date == null) {
                return "";
            }
            String stringFormatData = format.format(date);
            String[] gregorianDate = stringFormatData.split("-");
            JalaliCalendar jalaliCalendar = new JalaliCalendar(new GregorianCalendar(Integer.parseInt(gregorianDate[0]),
                    Integer.parseInt(gregorianDate[1]) - 1, Integer.parseInt(gregorianDate[2])));
            return jalaliCalendar.getDayOfWeekString();
        } catch (ParseException e) {
            e.printStackTrace();
        }
        return "";
    }

    public static JalaliCalendar getPersianDateJalali(String gregorianDateValue) {
        String[] dates = gregorianDateValue.split(" ");
        gregorianDateValue = dates[0];
        try {
            SimpleDateFormat format = new SimpleDateFormat(
                    "yyyy-MM-dd", Locale.US);
            Date date = format.parse(gregorianDateValue);
            if (date == null) {
                return null;
            }
            String stringFormatData = format.format(date);
            String[] gregorianDate = stringFormatData.split("-");
            return new JalaliCalendar(new GregorianCalendar(Integer.parseInt(gregorianDate[0]),
                    Integer.parseInt(gregorianDate[1]) - 1, Integer.parseInt(gregorianDate[2])));
        } catch (ParseException e) {
            e.printStackTrace();
        }
        return null;
    }

    public static JalaliCalendar getPersianDateJalali(int year, int month, int day) {
        return new JalaliCalendar(year, month, day);
    }


}

File: DashwoodCalendar/src/main/java/com/dashwood/dashwoodcalendar/handler/DateConverter.kt
package com.dashwood.dashwoodcalendar.handler

import ir.huri.jcal.JalaliCalendar
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.GregorianCalendar
import java.util.Locale

object DateConverter {

    fun persianDateToGregorianDate(persianDate: String): String {
        return try {
            val format = SimpleDateFormat("yyyy-MM-dd", Locale.US)
            val date: Date? = format.parse(persianDate)
            if (date == null) {
                ""
            } else {
                val persianFormatDate = format.format(date)
                val persianDates = persianFormatDate.split("-")
                val jalaliCalendar = JalaliCalendar(
                    persianDates[0].toInt(),
                    persianDates[1].toInt(),
                    persianDates[2].toInt()
                )
                val gregorianCalendar = jalaliCalendar.toGregorian()
                val day = gregorianCalendar[Calendar.DAY_OF_MONTH]
                val stringDay = if (day < 10) "0$day" else day.toString()
                stringDay + "-" +
                        (gregorianCalendar[Calendar.MONTH] + 1) + "-" +
                        gregorianCalendar[Calendar.YEAR]
            }
        } catch (e: ParseException) {
            e.printStackTrace()
            ""
        }
    }

    fun persianDateToGregorianDate(year: Int, month: Int, day: Int): String {
        val jalaliCalendar = JalaliCalendar(year, month, day)
        val gregorianCalendar = jalaliCalendar.toGregorian()
        val newDay = gregorianCalendar[Calendar.DAY_OF_MONTH]
        val stringDay = if (newDay < 10) "0$newDay" else newDay.toString()
        return gregorianCalendar[Calendar.YEAR].toString() + "-" +
                (gregorianCalendar[Calendar.MONTH] + 1) + "-" +
                stringDay
    }

    fun persianDateToGregorianDateFirstReturnYear(persianDate: String): String {
        return try {
            val format = SimpleDateFormat("yyyy-MM-dd", Locale.US)
            val date: Date? = format.parse(persianDate)
            if (date == null) {
                ""
            } else {
                val persianFormatDate = format.format(date)
                val persianDates = persianFormatDate.split("-")
                val jalaliCalendar = JalaliCalendar(
                    persianDates[0].toInt(),
                    persianDates[1].toInt(),
                    persianDates[2].toInt()
                )
                val gregorianCalendar = jalaliCalendar.toGregorian()
                val day = gregorianCalendar[Calendar.DAY_OF_MONTH]
                val stringDay = if (day < 10) "0$day" else day.toString()
                gregorianCalendar[Calendar.YEAR].toString() + "-" +
                        (gregorianCalendar[Calendar.MONTH] + 1) + "-" +
                        stringDay
            }
        } catch (e: ParseException) {
            e.printStackTrace()
            ""
        }
    }

    fun gregorianToPersianWithMonthStringName(gregorianDateValue: String): String {
        return try {
            val format = SimpleDateFormat("yyyy-MM-dd", Locale.US)
            val date: Date? = format.parse(gregorianDateValue)
            if (date == null) {
                ""
            } else {
                val stringFormatData = format.format(date)
                val gregorianDate = stringFormatData.split("-")
                val jalaliCalendar = JalaliCalendar(
                    GregorianCalendar(
                        gregorianDate[0].toInt(),
                        gregorianDate[1].toInt() - 1,
                        gregorianDate[2].toInt()
                    )
                )
                val day = jalaliCalendar.day
                val stringDay = if (day < 10) "0$day" else day.toString()
                "$stringDay ${jalaliCalendar.monthString} ${jalaliCalendar.year}"
            }
        } catch (e: ParseException) {
            e.printStackTrace()
            ""
        }
    }

    fun gregorianToPersian(gregorianDateValue: String): String {
        var value = gregorianDateValue
        val dates = value.split(" ")
        value = dates[0]
        return try {
            val format = SimpleDateFormat("yyyy-MM-dd", Locale.US)
            val date: Date? = format.parse(value)
            if (date == null) {
                ""
            } else {
                val stringFormatData = format.format(date)
                val gregorianDate = stringFormatData.split("-")
                val jalaliCalendar = JalaliCalendar(
                    GregorianCalendar(
                        gregorianDate[0].toInt(),
                        gregorianDate[1].toInt() - 1,
                        gregorianDate[2].toInt()
                    )
                )
                val day = jalaliCalendar.day
                val stringDay = if (day < 10) "0$day" else day.toString()
                "${jalaliCalendar.year}-${jalaliCalendar.month}-$stringDay"
            }
        } catch (e: ParseException) {
            e.printStackTrace()
            ""
        }
    }

    fun gregorianToPersian(year: Int, month: Int, day: Int): String {
        val jalaliCalendar = JalaliCalendar(GregorianCalendar(year, month, day))
        val newDay = jalaliCalendar.day
        val stringDay = if (newDay < 10) "0$newDay" else newDay.toString()
        return "${jalaliCalendar.year}-${jalaliCalendar.month}-$stringDay"
    }

    fun getMonthLength(gregorianDateValue: String): Int {
        var value = gregorianDateValue
        val dates = value.split(" ")
        value = dates[0]
        return try {
            val format = SimpleDateFormat("yyyy-MM-dd", Locale.US)
            val date: Date? = format.parse(value)
            if (date == null) {
                0
            } else {
                val stringFormatData = format.format(date)
                val gregorianDate = stringFormatData.split("-")
                val jalaliCalendar = JalaliCalendar(
                    GregorianCalendar(
                        gregorianDate[0].toInt(),
                        gregorianDate[1].toInt() - 1,
                        gregorianDate[2].toInt()
                    )
                )
                jalaliCalendar.monthLength
            }
        } catch (e: ParseException) {
            e.printStackTrace()
            0
        }
    }

    fun getMonthStartWeekName(gregorianDateValue: String): String {
        var value = gregorianDateValue
        val dates = value.split(" ")
        value = dates[0]
        return try {
            val format = SimpleDateFormat("yyyy-MM-dd", Locale.US)
            val date: Date? = format.parse(value)
            if (date == null) {
                ""
            } else {
                val stringFormatData = format.format(date)
                val gregorianDate = stringFormatData.split("-")
                val jalaliCalendar = JalaliCalendar(
                    GregorianCalendar(
                        gregorianDate[0].toInt(),
                        gregorianDate[1].toInt() - 1,
                        gregorianDate[2].toInt()
                    )
                )
                jalaliCalendar.dayOfWeekString
            }
        } catch (e: ParseException) {
            e.printStackTrace()
            ""
        }
    }

    fun getPersianDateJalali(gregorianDateValue: String): JalaliCalendar? {
        var value = gregorianDateValue
        val dates = value.split(" ")
        value = dates[0]
        return try {
            val format = SimpleDateFormat("yyyy-MM-dd", Locale.US)
            val date: Date? = format.parse(value)
            if (date == null) {
                null
            } else {
                val stringFormatData = format.format(date)
                val gregorianDate = stringFormatData.split("-")
                JalaliCalendar(
                    GregorianCalendar(
                        gregorianDate[0].toInt(),
                        gregorianDate[1].toInt() - 1,
                        gregorianDate[2].toInt()
                    )
                )
            }
        } catch (e: ParseException) {
            e.printStackTrace()
            null
        }
    }

    fun getPersianDateJalali(year: Int, month: Int, day: Int): JalaliCalendar {
        return JalaliCalendar(year, month, day)
    }

    fun persianDateToGregorianDateWithSlash(persianDate: String): String {
        return try {
            val format = SimpleDateFormat("yyyy/MM/dd", Locale.ROOT)
            val date: Date? = format.parse(persianDate)
            if (date == null) {
                ""
            } else {
                val persianFormatDate = format.format(date)
                val persianDates = persianFormatDate.split("/")
                val jalaliCalendar = JalaliCalendar(
                    persianDates[0].toInt(),
                    persianDates[1].toInt(),
                    persianDates[2].toInt()
                )
                val gregorianCalendar = jalaliCalendar.toGregorian()
                val day = gregorianCalendar[Calendar.DAY_OF_MONTH]
                val stringDay = if (day < 10) "0$day" else day.toString()
                val month = gregorianCalendar[Calendar.MONTH] + 1
                val stringMonth = if (month < 10) "0$month" else month.toString()
                "${gregorianCalendar[Calendar.YEAR]}/$stringMonth/$stringDay"
            }
        } catch (e: ParseException) {
            e.printStackTrace()
            ""
        }
    }

    fun gregorianToPersianWithSlash(gregorianDateValue: String): String {
        var value = gregorianDateValue
        val dates = value.split(" ")
        value = dates[0]
        return try {
            val format = SimpleDateFormat("yyyy/MM/dd", Locale.ROOT)
            val date: Date? = format.parse(value)
            if (date == null) {
                ""
            } else {
                val stringFormatData = format.format(date).replace("/", "-")
                val gregorianDate = stringFormatData.split("-")
                val jalaliCalendar = JalaliCalendar(
                    GregorianCalendar(
                        gregorianDate[0].toInt(),
                        gregorianDate[1].toInt() - 1,
                        gregorianDate[2].toInt()
                    )
                )
                val day = jalaliCalendar.day
                val stringDay = if (day < 10) "0$day" else day.toString()
                val month = jalaliCalendar.month
                val stringMonth = if (month < 10) "0$month" else month.toString()
                "${jalaliCalendar.year}/$stringMonth/$stringDay"
            }
        } catch (e: ParseException) {
            e.printStackTrace()
            ""
        }
    }
}
